controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java

```diff
@@ -0,0 +1,811 @@
+        return getEpochTransactionsTable(epoch)
+                .thenCompose(epochTable -> {
+                    return Futures.exceptionallyComposeExpecting(
+                            storeHelper.addNewEntry(scope, epochTable, txId.toString(), txnRecord),
+                            DATA_NOT_FOUND_PREDICATE, () -> createEpochTxnTableAndEntries(scope, epochTable, epoch, txId, txnRecord));
+                });
+    }
+
+    private CompletableFuture<Version> createEpochTxnTableAndEntries(String scope, String epochTable, int epoch, 
+                                                                     UUID txnId, byte[] txnRecord) {
+        return getEpochsWithTransactionsTable()
+                .thenCompose(epochWithTxnTable ->
+                        storeHelper.addNewEntryIfAbsent(scope, epochWithTxnTable, "" + epoch, new byte[0])
+                                   .thenCompose(added ->
+                                           storeHelper.createTable(scope, epochTable)
+                                                      .thenAccept(v -> log.debug("transactions in epoch {}/{} table created ", scope, epochTable)))
+                                   // Note: the epoch table could still have been deleted as we attempt to create transaction in it. 
+                                   // We want to create the table again but segment store does not allow us to recreate 
+                                   // the table with same name. So we are not deleting the epoch table when they are empty.  
+                                   .thenCompose(tableCreated -> createNewTransaction(epoch, txnId, txnRecord)));    
+    }
+    
+    @Override
+    CompletableFuture<Data> getActiveTx(final int epoch, final UUID txId) {
+        return getEpochTransactionsTable(epoch)
+                .thenCompose(epochTxnTable -> storeHelper.getEntry(getScope(), epochTxnTable, txId.toString()));
+    }
+
+    @Override
+    CompletableFuture<Version> updateActiveTx(final int epoch, final UUID txId, final Data data) {
+        return getEpochTransactionsTable(epoch)
+                .thenCompose(epochTxnTable -> storeHelper.updateEntry(getScope(), epochTxnTable, txId.toString(), data));
+    }
+
+    @Override
+    CompletableFuture<Void> removeActiveTxEntry(final int epoch, final UUID txId) {
+        // 1. remove txn from txn-in-epoch table
+        // 2. get current epoch --> if txn-epoch < activeEpoch.reference epoch, try deleting empty epoch table.
+        return getEpochTransactionsTable(epoch)
+                .thenCompose(epochTransactionsTableName -> 
+                        storeHelper.removeEntry(getScope(), epochTransactionsTableName, txId.toString()));
+        // We have commentted the following code because we cant recreate the table with the same name in segment store. 
+        // so we will not even remove it. So we have proliferation of tables. 
+        // .thenCompose(v -> tryRemoveEpoch(epoch)))
```

Comment:
commented out code

Reply:
I have added the reason for this.. 
The problem is -- we want to delete epoch specific tables where transactions are created once all transactions within it are removed and there is little chance of transactions being created (during scale, even if newer epochs are created some transaction creation attempt may happen on old epoch.. So ideally, i would want to remove empty table and let it get recreated if a transaction creation is attempted on this epoch. But we cant delete and recreate the tables.. 

So i have commented out the deletion for now.. for now, these tables will get deleted when a stream is sealed/deleted. 
but that would mean proliferation of tables.
I will mitigate this after opening a new issue. to mitigate this we will need to allow txn creation to fail and be retried on a different epoch if table does not exist and txn-epoch lags active-epoch.

Reply:
I changed the logic here.. now i dont group transactions by epochs. I keep all transactions in a single table. 
So i do not have to worry about the above scenario.